src/meta/processors/zone/DropHostsProcessor.cpp

```diff
@@ -0,0 +1,194 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License.
+ */
+
+#include "meta/processors/zone/DropHostsProcessor.h"
+
+namespace nebula {
+namespace meta {
+
+void DropHostsProcessor::process(const cpp2::DropHostsReq& req) {
+  folly::SharedMutex::WriteHolder zHolder(LockUtils::zoneLock());
+  folly::SharedMutex::WriteHolder mHolder(LockUtils::machineLock());
+  auto hosts = req.get_hosts();
+  if (std::unique(hosts.begin(), hosts.end()) != hosts.end()) {
+    LOG(ERROR) << "Hosts have duplicated element";
+    handleErrorCode(nebula::cpp2::ErrorCode::E_INVALID_PARM);
+    onFinished();
+    return;
+  }
+
+  if (hosts.empty()) {
+    LOG(ERROR) << "Hosts is empty";
+    handleErrorCode(nebula::cpp2::ErrorCode::E_INVALID_PARM);
+    onFinished();
+    return;
+  }
+
+  std::vector<std::string> data;
+  std::vector<kvstore::KV> rewriteData;
+  // Check that partition is not held on the host
+  const auto& spacePrefix = MetaKeyUtils::spacePrefix();
+  auto spaceIterRet = doPrefix(spacePrefix);
+  auto spaceIter = nebula::value(spaceIterRet).get();
+  nebula::cpp2::ErrorCode code = nebula::cpp2::ErrorCode::SUCCEEDED;
+  while (spaceIter->valid()) {
+    auto spaceId = MetaKeyUtils::spaceId(spaceIter->key());
+    auto spaceKey = MetaKeyUtils::spaceKey(spaceId);
+    auto ret = doGet(spaceKey);
+    if (!nebula::ok(ret)) {
+      code = nebula::error(ret);
+      LOG(ERROR) << "Get Space " << spaceId
+                 << " error: " << apache::thrift::util::enumNameSafe(code);
+      break;
+    }
+
+    const auto& partPrefix = MetaKeyUtils::partPrefix(spaceId);
+    auto partIterRet = doPrefix(partPrefix);
+    auto partIter = nebula::value(partIterRet).get();
+    while (partIter->valid()) {
+      auto partHosts = MetaKeyUtils::parsePartVal(partIter->val());
+      for (auto& h : partHosts) {
+        if (std::find(hosts.begin(), hosts.end(), h) != hosts.end()) {
+          LOG(ERROR) << h << " is related with partition";
+          code = nebula::cpp2::ErrorCode::E_CONFLICT;
+          break;
+        }
+      }
+      partIter->next();
+    }
+    spaceIter->next();
+  }
+
+  if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
+    handleErrorCode(code);
+    onFinished();
+    return;
+  }
+
+  // Remove points related to the zone
+  const auto& prefix = MetaKeyUtils::zonePrefix();
+  auto iterRet = doPrefix(prefix);
+  if (!nebula::ok(iterRet)) {
+    auto retCode = nebula::error(iterRet);
+    LOG(ERROR) << "List zones failed, error: " << apache::thrift::util::enumNameSafe(retCode);
+    handleErrorCode(retCode);
+    onFinished();
+    return;
+  }
+
+  auto iter = nebula::value(iterRet).get();
+  while (iter->valid()) {
+    auto zoneKey = iter->key();
+    auto hs = MetaKeyUtils::parseZoneHosts(iter->val());
```

Comment:
1. The if branch in line 86
    * Where do we check each host in `hs` has no related partitions?
    * If all hosts has no related partitions, the zone itself should be deleted
2. The else branch in 101
    * Where do we check each host in `hs` has no related partitions?